x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/PreAuthorizationUtil.java

```diff
@@ -0,0 +1,209 @@
+            // If role is null, it means a custom authorization engine is in use, hence we cannot do the optimization here.
+            return;
+        }
+
+        // We can't safely pre-authorize actions if DLS or FLS is configured without passing IAC as well with the authorization result.
+        // For simplicity, we only pre-authorize actions when FLS and DLS are not configured, otherwise we would have to compute and send
+        // indices access control as well, which we want to avoid with this optimization.
+        if (role.hasFieldOrDocumentLevelSecurity()) {
+            return;
+        }
+
+        final Optional<ParentActionAuthorization> existingParentAuthorization = Optional.ofNullable(
+            securityContext.getParentAuthorization()
+        );
+        if (existingParentAuthorization.isPresent()) {
+            if (existingParentAuthorization.get().action().equals(parentAction)) {
+                // Single request can fan-out a child action to multiple nodes in the cluster, e.g. node1 and node2.
+                // Sending a child action to node1 would have already put parent authorization in the thread context.
+                // To avoid attempting to pre-authorize the same parent action twice we simply return here
+                // since parent authorization is already set in the thread context.
+                if (logger.isDebugEnabled()) {
+                    logger.debug("authorization for parent action [" + parentAction + "] is already set in the thread context");
+                }
+            } else {
+                throw new AssertionError(
+                    "found parent authorization for action ["
+                        + existingParentAuthorization.get().action()
+                        + "] while attempting to set authorization for new parent action ["
+                        + parentAction
+                        + "]"
+                );
+            }
+        } else {
+            if (logger.isDebugEnabled()) {
+                logger.debug("adding authorization for parent action [" + parentAction + "] to the thread context");
+            }
+            securityContext.setParentAuthorization(new ParentActionAuthorization(parentAction));
+        }
+    }
+
+    private static boolean isChildActionWhitelistedForParent(final String parent, final String child) {
+        final Set<String> children = CHILD_ACTIONS_PRE_AUTHORIZED_BY_PARENT.get(parent);
+        return children != null && (parent.equals(child) || children.contains(child));
+    }
+
+    public static boolean shouldRemoveParentAuthorizationFromThreadContext(
+        Optional<String> remoteClusterAlias,
+        String childAction,
+        SecurityContext securityContext
+    ) {
+        final ParentActionAuthorization parentAuthorization = securityContext.getParentAuthorization();
+        if (parentAuthorization == null) {
+            // Nothing to remove.
+            return false;
+        }
```

Comment:
nit: better make the remove a noop if the header is not present, and remove this "nothing to remove" check.

Reply:
I'm not sure if I can get rid of this `null` check since I access it at line 155:
https://github.com/elastic/elasticsearch/blob/72f0951379d35ea66c314ad57c0da75bd0f43b62/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/PreAuthorizationUtil.java#L155

but I like the idea to make the remove method noop if the header is not set.